### PR TITLE
Add more caching to CRecoveredSigsDb and use salted hashing for externally provided keys

### PR DESCRIPTION
**Pull request by codablock**
See individual commits.

### CODE REVIEW

File: src/hash.cpp [hunk 1]
```diff
@@ -249,3 +250,7 @@ uint64_t SipHashUint256Extra(uint64_t k0, uint64_t k1, const uint256& val, uint3
     SIPROUND;
     return v0 ^ v1 ^ v2 ^ v3;
 }
+
+SaltedHasherBase::SaltedHasherBase() : k0(GetRand(std::numeric_limits<uint64_t>::max())), k1(GetRand(std::numeric_limits<uint64_t>::max())) {}
```
Comment (UdjinM6) [inline src/hash.cpp]: This thing doesn't belong here it seems, should be moved out of `consensus` stuff and into `util` or `common` somewhere.

Comment (codablock) [reply src/hash.cpp]: The salted hashers got their own source files now to avoid this problem.

Comment (UdjinM6) [reply src/hash.cpp]: 👍
Let's see if Travis is happy this time :)

**Comment by UdjinM6** [comment]
need rebase

**Comment by codablock** [comment]
Rebased and force-pushed

**Pull request merged**